### PR TITLE
Add PhpUnitInternalClassFixer

### LINKED ISSUE #3741: Feature request: Non abstract phpunit test classes should be final
```diff
- class MyTest extends \PHPUnit\Framework\TestCase {}
+ final class MyTest extends \PHPUnit\Framework\TestCase {}
```

It could replace the `testThatTestClassesAreTraitOrAbstractOrFinal` auto-review test.

### PR DESCRIPTION
**Pull request by BackEndTea**
Fixes #3741 

Adds `@internal` annotation to phpunit test classes. This fixes #3741 if used in combination with the `final_internal_class` fixer.

### CODE REVIEW

**Comment by keradus** [comment]
Looks promising, super :+1: for working on this @BackEndTea ;)

Please add this rule to local `.php_cs.dist` file

(review is still needed)

**Comment by BackEndTea** [comment]
> Please add this rule to local .php_cs.dist file

I did that, but it 'fixes' some classes that aren't meant to be internal.

**Comment by keradus** [comment]
Then it shows some shortcomings of the implementation

File: src/Fixer/PhpUnit/PhpUnitInternalClassFixer.php [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+    /**
+     * @param Tokens $tokens
+     *
+     * @return \Generator <int, int>
+     */
+    private function findPhpUnitClasses(Tokens $tokens)
+    {
+        $phpUnitTestCaseIndicator = new PhpUnitTestCaseIndicator();
+
+        for ($index = $tokens->count() - 1; $index >= 0; --$index) {
```
Comment (SpacePossum) [inline src/Fixer/PhpUnit/PhpUnitInternalClassFixer.php]: could be simplified after https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3451

File: src/Fixer/PhpUnit/PhpUnitInternalClassFixer.php [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        return new FixerDefinition(
+            'All PHPUnit test classes should be marked as internal.',
+            [new CodeSample("<?php class MyTest extends TestCase {}\n")]
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isCandidate(Tokens $tokens)
```
Comment (SpacePossum) [inline src/Fixer/PhpUnit/PhpUnitInternalClassFixer.php]: please move below `getPriority`

File: README.rst [hunk 3]
```diff
@@ -1121,6 +1121,10 @@ Choose from the list of available rules:
 
   PHPUnit annotations should be a FQCNs including a root namespace.
 
+* **php_unit_internal_class**
+
+  All PHPUnit test classes should be marked as internal.
```
Comment (keradus) [inline README.rst]: I would either not mark with `@internal` phpunit classes that are abstract, or add config option to control it.
Abstract `*TestCase` are often use as extension point for external ppl, same like in our repo.

File: src/Fixer/PhpUnit/PhpUnitInternalClassFixer.php (hunk 2 above)
Comment (BackEndTea) [reply src/Fixer/PhpUnit/PhpUnitInternalClassFixer.php]: Done

File: README.rst (hunk 3 above)
Comment (BackEndTea) [reply README.rst]: I've added the options
- `normal`
- `final`
- `abstract`

Normal meaning non final, non abstract.

File: src/Fixer/PhpUnit/PhpUnitInternalClassFixer.php [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+                ->getOption(),
+        ]);
+    }
+
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        $phpUnitTestCaseIndicator = new PhpUnitTestCaseIndicator();
+
+        foreach ($phpUnitTestCaseIndicator->findPhpUnitClasses($tokens) as $indexes) {
+            $this->markClassInternal($tokens, $indexes[0]);
```
[the remark this answers is not in the document]
Comment (julienfalque) [reply src/Fixer/PhpUnit/PhpUnitInternalClassFixer.php]: Like you said, it's an implementation detail and could change easily, I'm kinda :-1: to rely on it.

[the remark this answers is not in the document]
Comment (keradus) [reply src/Fixer/PhpUnit/PhpUnitInternalClassFixer.php]: @BackEndTea , if you would iterate backwards, we would be able to merge this way faster, without having alignment on relying on implementation detail such this one, especially when whole PHP CS Fixer is not yet relying on generators

**Comment by keradus** [comment]
@BackEndTea , can you adjust tets as well?

File: src/Indicator/PhpUnitTestCaseIndicator.php [hunk 5]
```diff
@@ -51,24 +51,54 @@ public function isPhpUnitClass(Tokens $tokens, $index)
...
     /**
      * @param Tokens $tokens
+     * @param bool   $beginAtBottom whether we should start yielding phpunit classes from the bottom of the file
      *
      * @return \Generator array of [int start, int end] indexes from sooner to later classes
      */
-    public function findPhpUnitClasses(Tokens $tokens)
+    public function findPhpUnitClasses(Tokens $tokens, $beginAtBottom = true)
+    {
+        if (!$beginAtBottom) {
```
Comment (SpacePossum) [inline src/Indicator/PhpUnitTestCaseIndicator.php]:
```php
        return $beginAtBottom 
            ? $this->findPhpUnitClassesStartingAtBottom($tokens)
            : $this->findPhpUnitClassesStartingFromTop($tokens)
        ;
```

ditches out the not needed `!`

File: src/Indicator/PhpUnitTestCaseIndicator.php [hunk 6]
```diff
@@ -51,24 +51,54 @@ public function isPhpUnitClass(Tokens $tokens, $index)
...
+            if (!$tokens[$index]->isGivenKind(T_CLASS) || !$this->isPhpUnitClass($tokens, $index)) {
+                continue;
+            }
+
+            $startIndex = $tokens->getNextTokenOfKind($index, ['{']);
+            $endIndex = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_CURLY_BRACE, $startIndex);
+            yield [$startIndex, $endIndex];
+        }
+    }
 }
```
Comment (SpacePossum) [inline src/Indicator/PhpUnitTestCaseIndicator.php]: I think `findPhpUnitClassesStartingFromTop` and `findPhpUnitClassesStartingAtBottom` can unified into something like;

```php
    /**
     * @param Tokens $tokens
     * @param int    $direction 1|-1
     *
     * @return \Generator
     */
    private function findPhpUnitClasses(Tokens $tokens, $direction)
    {
        $index = 1 === $direction ? 0 : $tokens->count() - 1;
        while (true) {
            if (!$tokens[$index]->isGivenKind(T_CLASS) || !$this->isPhpUnitClass($tokens, $index)) {
                continue;
            }

            $startIndex = $tokens->getTokenOfKindSibling($index, $direction, ['{'], false);
            if (null === $startIndex) {
                return;
            }

            $endIndex = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_CURLY_BRACE, $startIndex);
            yield [$startIndex, $endIndex];

            $index += $direction;
        }
    }
```

Comment (keradus) [reply src/Indicator/PhpUnitTestCaseIndicator.php]: almost,
`$tokens->getTokenOfKindSibling($index, $direction, ['{'])` is wrong, as `{` will always be after `T_CLASS`, never before.

I will commit the update to the PR

Comment (keradus) [reply src/Indicator/PhpUnitTestCaseIndicator.php]: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3745/commits/d4c609323be6bc82b084d5511be81f1e2991fef0

Comment (keradus) [reply src/Indicator/PhpUnitTestCaseIndicator.php]:
```php
    public function findPhpUnitClasses(Tokens $tokens, $beginAtBottom = true)
    {
        $direction = $beginAtBottom ? -1 : 1;
        for ($index = 1 === $direction ? 0 : $tokens->count() - 1; $tokens->offsetExists($index); $index += $direction) {
            if (!$tokens[$index]->isGivenKind(T_CLASS) || !$this->isPhpUnitClass($tokens, $index)) {
                continue;
            }
            $startIndex = $tokens->getNextTokenOfKind($index, ['{'], false);
            if (null === $startIndex) {
                return;
            }
            $endIndex = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_CURLY_BRACE, $startIndex);
            yield [$startIndex, $endIndex];
        }
    }
```

**Comment by keradus** [comment]
Thank you @BackEndTea.

File: src/Indicator/PhpUnitTestCaseIndicator.php (hunk 6 above)
Comment (SpacePossum) [reply src/Indicator/PhpUnitTestCaseIndicator.php]: thanks @keradus :+1:

**Comment by SpacePossum** [comment]
thanks @BackEndTea : )

**Pull request merged**